#!/usr/bin/env python3
"""
è¯¾è¡¨æ¨¡å— API æµ‹è¯•è„šæœ¬ï¼ˆæ”¯æŒ JWT è®¤è¯ï¼‰
ä½¿ç”¨æ–¹æ³•ï¼š
1. å®‰è£…ä¾èµ–ï¼špip install requests protobuf
2. ç¼–è¯‘ protoï¼šprotoc --python_out=. proto/course.proto
3. è¿è¡Œæµ‹è¯•ï¼špython test_api.py
"""

import requests
import sys
import os

# æ·»åŠ ç”Ÿæˆçš„ proto æ–‡ä»¶è·¯å¾„
sys.path.insert(0, os.path.dirname(__file__))

try:
    from proto import course_pb2
    from generate_token import generate_jwt_token
except ImportError as e:
    print(f"é”™è¯¯ï¼šæ‰¾ä¸åˆ°å¿…è¦æ¨¡å—: {e}")
    print("è¯·ç¡®ä¿ï¼š")
    print("1. è¿è¡Œï¼šprotoc --python_out=. proto/course.proto")
    print("2. generate_token.py æ–‡ä»¶å­˜åœ¨")
    sys.exit(1)

BASE_URL = "http://localhost:3000"

# JWT è®¤è¯é…ç½®
TEST_USER_ID = 1  # æµ‹è¯•ç”¨æˆ· IDï¼Œå¯ä»¥é€šè¿‡ç¯å¢ƒå˜é‡æˆ–å‘½ä»¤è¡Œå‚æ•°ä¿®æ”¹
JWT_TOKEN = None  # å…¨å±€ token å˜é‡

def set_test_user(user_id):
    """åˆ‡æ¢æµ‹è¯•ç”¨æˆ·"""
    global TEST_USER_ID, JWT_TOKEN
    TEST_USER_ID = user_id
    JWT_TOKEN = None  # é‡ç½® tokenï¼Œä¸‹æ¬¡è¯·æ±‚æ—¶ä¼šé‡æ–°ç”Ÿæˆ
    print(f"ğŸ”„ åˆ‡æ¢åˆ°ç”¨æˆ· ID: {user_id}")

def get_auth_headers():
    """è·å–å¸¦è®¤è¯çš„è¯·æ±‚å¤´"""
    global JWT_TOKEN
    if JWT_TOKEN is None:
        try:
            JWT_TOKEN = generate_jwt_token(TEST_USER_ID)
            print(f"ğŸ”‘ ä¸ºç”¨æˆ· {TEST_USER_ID} ç”Ÿæˆ JWT Token")
        except Exception as e:
            print(f"âŒ ç”Ÿæˆ JWT Token å¤±è´¥: {e}")
            return {}
    
    return {
        'Authorization': f'Bearer {JWT_TOKEN}',
        'Accept': 'application/x-protobuf'
    }

def get_public_headers():
    """è·å–å…¬å¼€æ¥å£çš„è¯·æ±‚å¤´ï¼ˆæ— éœ€è®¤è¯ï¼‰"""
    return {'Accept': 'application/x-protobuf'}

def print_separator(title):
    print("\n" + "="*60)
    print(f"  {title}")
    print("="*60)

def test_get_semesters():
    """æµ‹è¯•è·å–å­¦æœŸåˆ—è¡¨ï¼ˆå…¬å¼€æ¥å£ï¼‰"""
    print_separator("æµ‹è¯•ï¼šè·å–å­¦æœŸåˆ—è¡¨ï¼ˆå…¬å¼€æ¥å£ï¼‰")
    
    response = requests.get(
        f"{BASE_URL}/api/v1/semesters",
        headers=get_public_headers()
    )
    
    print(f"çŠ¶æ€ç : {response.status_code}")
    print(f"Content-Type: {response.headers.get('Content-Type')}")
    
    if response.status_code == 200:
        result = course_pb2.GetSemestersResponse()
        result.ParseFromString(response.content)
        
        print(f"\nå“åº”ç : {result.code}")
        print(f"æ¶ˆæ¯: {result.message}")
        print(f"\nå­¦æœŸåˆ—è¡¨ (å…± {len(result.data.semesters)} ä¸ª):")
        for semester in result.data.semesters:
            current = "âœ“ å½“å‰å­¦æœŸ" if semester.is_current else ""
            print(f"  - ID: {semester.id}, åç§°: {semester.name} {current}")
            print(f"    æ—¶é—´: {semester.start_date} ~ {semester.end_date}")
    else:
        print(f"è¯·æ±‚å¤±è´¥: {response.text}")

def test_get_public_courses(semester_id=None, page=1, page_size=5):
    """æµ‹è¯•è·å–å…¨æ ¡è¯¾ç¨‹ï¼ˆå…¬å¼€æ¥å£ï¼‰"""
    print_separator("æµ‹è¯•ï¼šè·å–å…¨æ ¡è¯¾ç¨‹ï¼ˆå…¬å¼€æ¥å£ï¼‰")
    
    params = {
        'page': page,
        'pageSize': page_size
    }
    if semester_id:
        params['semester_id'] = semester_id
    
    response = requests.get(
        f"{BASE_URL}/api/v1/courses",
        params=params,
        headers=get_public_headers()
    )
    
    print(f"çŠ¶æ€ç : {response.status_code}")
    
    if response.status_code == 200:
        result = course_pb2.GetPublicCoursesResponse()
        result.ParseFromString(response.content)
        
        print(f"\nå“åº”ç : {result.code}")
        print(f"æ¶ˆæ¯: {result.message}")
        print(f"\nåˆ†é¡µä¿¡æ¯:")
        print(f"  æ€»æ•°: {result.data.pagination.total}")
        print(f"  å½“å‰é¡µ: {result.data.pagination.page}/{result.data.pagination.pages}")
        print(f"\nè¯¾ç¨‹åˆ—è¡¨:")
        for course in result.data.list:
            print(f"  - {course.course_name} ({course.teacher_name})")
            print(f"    æ—¶é—´: å‘¨{course.day_of_week} ç¬¬{course.start_section}-{course.end_section}èŠ‚")
            print(f"    åœ°ç‚¹: {course.location}")
    else:
        print(f"è¯·æ±‚å¤±è´¥: {response.text}")

def test_get_schedule(semester_id, week=None):
    """æµ‹è¯•è·å–ç”¨æˆ·è¯¾è¡¨ï¼ˆéœ€è¦è®¤è¯ï¼‰"""
    print_separator("æµ‹è¯•ï¼šè·å–ç”¨æˆ·è¯¾è¡¨ï¼ˆéœ€è¦è®¤è¯ï¼‰")
    
    params = {'semester_id': semester_id}
    if week:
        params['week'] = week
    
    response = requests.get(
        f"{BASE_URL}/api/v1/schedule",
        params=params,
        headers=get_auth_headers()
    )
    
    print(f"çŠ¶æ€ç : {response.status_code}")
    
    if response.status_code == 200:
        result = course_pb2.GetScheduleResponse()
        result.ParseFromString(response.content)
        
        print(f"\nå“åº”ç : {result.code}")
        print(f"æ¶ˆæ¯: {result.message}")
        print(f"\nè¯¾è¡¨é¡¹ (å…± {len(result.data.items)} é¡¹):")
        for item in result.data.items:
            custom = "âœ“ è‡ªå®šä¹‰" if item.is_custom else ""
            print(f"  - {item.course_name} ({item.teacher_name}) {custom}")
            print(f"    æ—¶é—´: å‘¨{item.day_of_week} ç¬¬{item.start_section}-{item.end_section}èŠ‚")
            print(f"    åœ°ç‚¹: {item.location}")
            if item.color_hex:
                print(f"    é¢œè‰²: {item.color_hex}")
    else:
        print(f"è¯·æ±‚å¤±è´¥: {response.text}")

def test_add_schedule_items(semester_id):
    """æµ‹è¯•æ·»åŠ è¯¾è¡¨é¡¹ï¼ˆéœ€è¦è®¤è¯ï¼‰"""
    print_separator("æµ‹è¯•ï¼šæ·»åŠ è¯¾è¡¨é¡¹ï¼ˆéœ€è¦è®¤è¯ï¼‰")
    
    # æ„é€ è¯·æ±‚
    request = course_pb2.AddScheduleItemsRequest(
        semester_id=semester_id,
        items=[
            course_pb2.ScheduleItemInput(
                source_id=1,                   
                course_name="é«˜ç­‰æ•°å­¦A",
                teacher_name="å¼ æ•™æˆ",
                location="æ•™å­¦æ¥¼A-101",
                day_of_week=1,
                start_section=1,
                end_section=2,
                weeks=[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16],
                type="compulsory",              
                credits=4,
                description="é«˜ç­‰æ•°å­¦åŸºç¡€è¯¾ç¨‹",
                color_hex="#000000",            # é»˜è®¤é¢œè‰²ï¼ˆå¯é€‰ï¼‰
                is_custom=False                 # é»˜è®¤éè‡ªå®šä¹‰
            ),
            course_pb2.ScheduleItemInput(
                course_name="è‡ªå®šä¹‰è¯¾ç¨‹B",
                teacher_name="æè€å¸ˆ",
                location="å›¾ä¹¦é¦†",
                day_of_week=6,
                start_section=5,
                end_section=6,
                weeks=[10],
                type="é€‰ä¿®",
                credits=2,
                color_hex="#33FF57",
                is_custom=True
            ),
            course_pb2.ScheduleItemInput(
                course_name="æµ‹è¯•è¯¾ç¨‹B",  # æ–°è¯¾ç¨‹åç§°
                teacher_name="æè€å¸ˆ",  # å‡è®¾æ˜¯å¦ä¸€ä½æ•™å¸ˆ
                location="æ•™å­¦æ¥¼A101",  # åŒä¸€åœ°ç‚¹
                day_of_week=1,  # æ˜ŸæœŸä¸€
                start_section=1,  # å¼€å§‹èŠ‚æ¬¡ç›¸åŒ
                end_section=2,  # ç»“æŸèŠ‚æ¬¡ç›¸åŒ
                weeks=[8, 9, 11, 12, 13, 14],  # ä¸é‡åˆçš„å‘¨æ•°ç»„
                type="é€‰ä¿®",  # æˆ–è€…"å¿…ä¿®"ï¼Œæ ¹æ®å®é™…éœ€æ±‚
                credits=2,  # å­¦åˆ†å¯ä»¥æ ¹æ®å®é™…æƒ…å†µè°ƒæ•´
                description="è¿™æ˜¯å¦ä¸€ä¸ªæµ‹è¯•è¯¾ç¨‹",
                color_hex="#33FF57",  # ä¸åŒé¢œè‰²ä»¥åŒºåˆ†
                is_custom=True
            )
        ]
    )
    
    # è·å–è®¤è¯å¤´å¹¶æ·»åŠ  Content-Type
    headers = get_auth_headers()
    headers['Content-Type'] = 'application/x-protobuf'
    
    response = requests.post(
        f"{BASE_URL}/api/v1/schedule",
        headers=headers,
        data=request.SerializeToString()
    )
    
    print(f"çŠ¶æ€ç : {response.status_code}")
    
    if response.status_code == 200:
        result = course_pb2.AddScheduleItemsResponse()
        result.ParseFromString(response.content)
        
        print(f"\nå“åº”ç : {result.code}")
        print(f"æ¶ˆæ¯: {result.message}")
        print(f"\næˆåŠŸæ·»åŠ  {len(result.data.successful_items)} é¡¹:")
        for item in result.data.successful_items:
            print(f"  - ID: {item.id}, è¯¾ç¨‹: {item.course_name}")
        
        if result.data.failed_items:
            print(f"\nå¤±è´¥ {len(result.data.failed_items)} é¡¹:")
            for item in result.data.failed_items:
                print(f"  - è¯¾ç¨‹: {item.course_name}, é”™è¯¯: {item.error_message}")
    else:
        print(f"è¯·æ±‚å¤±è´¥: {response.text}")

def test_update_schedule_item(item_id):
    """æµ‹è¯•æ›´æ–°è¯¾è¡¨é¡¹ï¼ˆéœ€è¦è®¤è¯ï¼‰"""
    print_separator("æµ‹è¯•ï¼šæ›´æ–°è¯¾è¡¨é¡¹ï¼ˆéœ€è¦è®¤è¯ï¼‰")
    
    request = course_pb2.UpdateScheduleItemRequest(
        course_name="æ›´æ–°åçš„è¯¾ç¨‹å",
        teacher_name="ç‹è€å¸ˆ",
        location="æ–°æ•™å­¦æ¥¼B202",
        day_of_week=2,
        start_section=3,
        end_section=4,
        weeks=[1,2,3,4,5,6,7,8,10],
        type="å¿…ä¿®",
        credits=4,
        description="è¯¾ç¨‹å·²æ›´æ–°",
        color_hex="#3357FF"
    )
    
    # è·å–è®¤è¯å¤´å¹¶æ·»åŠ  Content-Type
    headers = get_auth_headers()
    headers['Content-Type'] = 'application/x-protobuf'
    
    response = requests.patch(
        f"{BASE_URL}/api/v1/schedule",
        params={'item_id': item_id},
        headers=headers,
        data=request.SerializeToString()
    )
    
    print(f"çŠ¶æ€ç : {response.status_code}")
    
    if response.status_code == 200:
        result = course_pb2.UpdateScheduleItemResponse()
        result.ParseFromString(response.content)
        
        print(f"\nå“åº”ç : {result.code}")
        print(f"æ¶ˆæ¯: {result.message}")
        print(f"\næ›´æ–°åçš„è¯¾è¡¨é¡¹:")
        item = result.data.item
        print(f"  - ID: {item.id}")
        print(f"  - è¯¾ç¨‹: {item.course_name} ({item.teacher_name})")
        print(f"  - æ—¶é—´: å‘¨{item.day_of_week} ç¬¬{item.start_section}-{item.end_section}èŠ‚")
    else:
        print(f"è¯·æ±‚å¤±è´¥: {response.text}")

def test_delete_schedule_item(item_id):
    """æµ‹è¯•åˆ é™¤è¯¾è¡¨é¡¹ï¼ˆéœ€è¦è®¤è¯ï¼‰"""
    print_separator("æµ‹è¯•ï¼šåˆ é™¤è¯¾è¡¨é¡¹ï¼ˆéœ€è¦è®¤è¯ï¼‰")
    
    response = requests.delete(
        f"{BASE_URL}/api/v1/schedule",
        params={'item_id': item_id},
        headers=get_auth_headers()
    )
    
    print(f"çŠ¶æ€ç : {response.status_code}")
    
    if response.status_code == 200:
        result = course_pb2.DeleteScheduleItemResponse()
        result.ParseFromString(response.content)
        
        print(f"\nå“åº”ç : {result.code}")
        print(f"æ¶ˆæ¯: {result.message}")
    else:
        print(f"è¯·æ±‚å¤±è´¥: {response.text}")

def test_unauthorized_access():
    """æµ‹è¯•æ— è®¤è¯è®¿é—®ï¼ˆåº”è¯¥å¤±è´¥ï¼‰"""
    print_separator("æµ‹è¯•ï¼šæ— è®¤è¯è®¿é—®ï¼ˆåº”è¯¥è¿”å› 401ï¼‰")
    
    response = requests.get(
        f"{BASE_URL}/api/v1/schedule",
        params={'semester_id': 1},
        headers={'Accept': 'application/x-protobuf'}  # ä¸åŒ…å« Authorization å¤´
    )
    
    print(f"çŠ¶æ€ç : {response.status_code}")
    if response.status_code == 401:
        print("âœ… æ­£ç¡®ï¼šæœªè®¤è¯è¯·æ±‚è¢«æ‹’ç»")
    else:
        print("âŒ é”™è¯¯ï¼šæœªè®¤è¯è¯·æ±‚åº”è¯¥è¿”å› 401")
    
    if response.content:
        print(f"å“åº”å†…å®¹: {response.content}")

def main():
    print("\n" + "ğŸ“ è¯¾è¡¨æ¨¡å— API æµ‹è¯•å·¥å…·ï¼ˆJWT è®¤è¯ç‰ˆï¼‰".center(60))
    
    # æ£€æŸ¥å‘½ä»¤è¡Œå‚æ•°
    if len(sys.argv) > 1:
        if sys.argv[1] in ['-h', '--help', 'help']:
            print("\nä½¿ç”¨æ–¹æ³•:")
            print("  python test_api.py [user_id]")
            print("\nå‚æ•°:")
            print("  user_id    æµ‹è¯•ç”¨æˆ·çš„ IDï¼ˆé»˜è®¤: 1ï¼‰")
            print("\nç¤ºä¾‹:")
            print("  python test_api.py        # ä½¿ç”¨é»˜è®¤ç”¨æˆ· ID 1")
            print("  python test_api.py 2      # ä½¿ç”¨ç”¨æˆ· ID 2")
            print("  python test_api.py 999    # ä½¿ç”¨ç”¨æˆ· ID 999")
            return
        
        try:
            user_id = int(sys.argv[1])
            set_test_user(user_id)
        except ValueError:
            print(f"âŒ æ— æ•ˆçš„ç”¨æˆ· ID: {sys.argv[1]}")
            print("ä½¿ç”¨æ–¹æ³•: python test_api.py [user_id]")
            print("è¿è¡Œ 'python test_api.py --help' æŸ¥çœ‹å¸®åŠ©")
            return
    
    try:
        # 0. æµ‹è¯•æ— è®¤è¯è®¿é—®
        test_unauthorized_access()
        
        # 1. è·å–å­¦æœŸåˆ—è¡¨ï¼ˆå…¬å¼€æ¥å£ï¼‰
        test_get_semesters()
        
        # 2. è·å–å…¨æ ¡è¯¾ç¨‹ï¼ˆå…¬å¼€æ¥å£ï¼‰
        test_get_public_courses(page=1, page_size=3)
        
        # 3. è·å–ç”¨æˆ·è¯¾è¡¨ï¼ˆéœ€è¦è®¤è¯ï¼‰
        test_get_schedule(semester_id=1)
        
        # 4. æ·»åŠ è¯¾è¡¨é¡¹ï¼ˆéœ€è¦è®¤è¯ï¼‰
        test_add_schedule_items(semester_id=1)

        # 4.1 å†æ¬¡è·å–è¯¾è¡¨ï¼ˆéœ€è¦è®¤è¯ï¼‰
        test_get_schedule(semester_id=1)
        
        # 5. æ›´æ–°è¯¾è¡¨é¡¹ï¼ˆéœ€è¦è®¤è¯ï¼Œéœ€è¦å…ˆæœ‰æ•°æ®ï¼‰
        test_update_schedule_item(item_id=2)
        
        # 6. åˆ é™¤è¯¾è¡¨é¡¹ï¼ˆéœ€è¦è®¤è¯ï¼Œéœ€è¦å…ˆæœ‰æ•°æ®ï¼‰
        test_delete_schedule_item(item_id=2)
        
        print_separator("æµ‹è¯•å®Œæˆ")
        print(f"âœ… ä½¿ç”¨çš„æµ‹è¯•ç”¨æˆ· ID: {TEST_USER_ID}")
        print(f"ğŸ”‘ JWT Token: {JWT_TOKEN[:50]}..." if JWT_TOKEN else "âŒ æœªç”Ÿæˆ Token")
        
    except requests.exceptions.ConnectionError:
        print("\nâŒ é”™è¯¯ï¼šæ— æ³•è¿æ¥åˆ°æœåŠ¡å™¨")
        print("è¯·ç¡®ä¿æœåŠ¡å™¨æ­£åœ¨è¿è¡Œï¼šcargo run")
    except Exception as e:
        print(f"\nâŒ é”™è¯¯ï¼š{e}")
        import traceback
        traceback.print_exc()

if __name__ == "__main__":
    main()
